File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
    candidates.sort((a, b) => a - b);
    const dfs = function (remSum, start, path) {
        if (remSum === 0) {
            result.push([...path]);
            return;
        }

        if (remSum < 0) return;

        for (let i = start; i < candidates.length; i++) {
            if (i > start && candidates[i - 1] === candidates[i]) continue;
            path.push(candidates[i]);
            dfs(remSum - candidates[i], i + 1, path);
            path.pop();
        }

    }

    const result = [];
    dfs(target, 0, []);

    return result;
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(remSum, start, path):
            if remSum == 0:
                result.append(path[:])

            if remSum <= 0:
                return

            for i in range(start, len(candidates)):
                if i > start and candidates[i - 1] == candidates[i]:
                    continue
                path.append(candidates[i])
                dfs(remSum - candidates[i], i + 1, path)
                path.pop()

        candidates.sort()
        result = []
        dfs(target, 0, [])

        return result
